Modules/_hashopenssl.c

```diff
@@ -1757,22 +1754,26 @@ _openssl_hash_name_mapper(const EVP_MD *md, const char *from,
 
 
 /* Ask OpenSSL for a list of supported ciphers, filling in a Python set. */
-static PyObject*
-generate_hash_name_list(void)
+static int
+hashlib_md_meth_names(PyObject *module)
 {
     _InternalNameMapperState state;
     state.set = PyFrozenSet_New(NULL);
     if (state.set == NULL)
-        return NULL;
+        return -1;
     state.error = 0;
 
     EVP_MD_do_all(&_openssl_hash_name_mapper, &state);
 
     if (state.error) {
         Py_DECREF(state.set);
-        return NULL;
+        return -1;
+    }
+    if (PyModule_AddObject(module, "openssl_md_meth_names", state.set)) {
+        Py_DECREF(state.set);
+        return -1;
     }
```

Comment:
IMO Py_DECREF(state.set); is also needed in case of success, no?

Reply:
fixed

Reply:
Oh wait. I was reading PyModule_AddType(), but this function uses the evil PyModule_AddObject() which has a very weird API: it **steals** a strong reference on success... `Py_DECREF(state.set);` must only be called on failure, not on success. Sorry :-(

Reply:
No? No! It's required. ``PyModule_AddObject()`` does not incref. In fact you have to incref objects before passing it to the function:

```
    Py_INCREF(ErrorObject);
    PyModule_AddObject(m, "error", ErrorObject);
```